### PR TITLE
fix(Scripts/Spells): Finger of Frost Shatter Combo.

### LINKED ISSUE #11338: [Mage][Talent] Finger of Frost / Ghost charge (FoF shatter combo)
### Current Behaviour

- Ghost charge not working.

### Expected Blizzlike Behaviour

- Should be able to weave in a 3rd spell on a FoF proc which you can clearly see on this [video](https://youtu.be/8ET2Ct9_Ma8?t=117), slow it down to 0.5 and watch the mage do fb (proc) > fb > fb > deep freeze combo on a single proc. Mentioned this on my previous post https://github.com/chromiecraft/chromiecraft/issues/2537 . I Think the way to make it work is to add a small delay to both the proc and consume which seems to be the case if you pay attention to the video ( This might be how procs worked in retaiI in general or this could all be purely a latency thing but I see this same "delay" as well when [Arcane concentration](https://wowgaming.altervista.org/aowow/?spell=12577) procs in other videos). Either way adding like a 300-500ms delay to both the proc and consume should fix this while also preventing issues like a consecutive FoF charge being consumed because of this "delay". It just has to be less than 1 second so you don't run into rare high haste occasion where ppl can weave in 2 abilities in that gap.

### Source

https://www.mmo-champion.com/threads/641024-FrostBite-Fingers-of-Frost-question
>Also, if you have FoF, you can actually cast 2 Frostbolts, the last one followed up with an Ice Lance - because that casting the Ice Lance instantaneously after the Frostbolt, you can practically achieve the shatter bonus three times, creating a "third" FoF charge.

>No, the charge is consumed on spell cast. The reason you can fit an Ice Lance in is network lag.
The FoF Shatter is basically a bug, as it goes directly against the talent tooltip that clearly states "your next 2 spells cast".
A normal Shatter combo relies on travel time, a game mechanic, to work, but this one can be impossible to pull to a player with the wrong amount of latency. Hence, if Blizzard intends for our DPS to be balanced around this, they have to make it more clear. The people over at EJ (Lhivera especially) are considerably vocal against this mechanic.

https://www.mmo-champion.com/threads/710176-Ghost-fingers-of-frost

>It works the same way as queueing spells, you get a FoF proc, frostbolt, frostbolt->ice lance. It actually works better on frostbite/novas, as you can FB-IL-IL if you have enough haste and are far enough away. It's pretty cute.
The idea is to cast it while you still have the buff or they still have the debuff. Latency is a factor, but you don't need particularly low latency, just not like, 800.

https://www.youtube.com/watch?v=8ET2Ct9_Ma8 Throughout this video you can see examples of everything I described happening including Frostbite always proccing with FoF. This is noticeable because every time you see the "Deep Freeze" yellow text (which represents a FoF proc) you also see an "immune" on the boss since you can't freeze bosses.

>A mage can use a "third" charge due to network latency. This is faciliated by using a spell with a [cast time](https://wowpedia.fandom.com/wiki/Casting_speed) greater than then [GCD](https://wowpedia.fandom.com/wiki/GCD) to consume the second charge, and then using an instant cast spell [Deep Freeze] or  [Ice Lance] are the best options for this) right as the spell finishes casting.

2 video sources were provided by [this comment in the CC Discord](https://discord.com/channels/788063059926712341/825732076975554571/958413698047807598): [1](https://www.youtube.com/watch?v=4X-GB9GB7Ic) and [2](https://www.youtube.com/watch?v=lr-ghl-CEL8)

[Another Post](https://www.mmo-champion.com/threads/704461-3-3-3-Mage-PvE-Guide-to-Frost-Basics)
>What ghosting FoF means is you can use an instant cast directly after using the second charge of your FoF proc, effectively applying 3 charges to your spells. Doing so will increase the crit chance of all spells cast by 50% and allow for you to use DF, your 30 second atomic bomb.]

Edit:
>Since Frostfire Bolt benefits from Ice Shards and Shatter, it is ideal to use its instant-castability immediately after the second Frostbolt during a Fingers of Frost proc.  Since the last Fingers of Frost charge is not consumed until the second spell hits the target, both the Frostbolt and the Frostfire Bolt will benefit from Fingers of Frost, resulting in what is commonly known as a Shatter Combo. ([source](https://frostisthenewblack.wordpress.com/2010/01/19/3-3-frost-mage-raiding-cheat-sheet/))

### Steps to reproduce the problem

1. Go to a dummy with only Fingers of Frost specced,
2. When at only 1 charge of Fingers of Frost, attempt (and fail) to cast Deep Freeze.

### Extra Notes

Originally Reported: https://github.com/chromiecraft/chromiecraft/issues/3123

Edit: Comment made by the original poster [bonfax4](https://github.com/bonfax4):

> [Spell batching guy you linked in discord](https://discord.com/channels/788063059926712341/825732076975554571/958502106191503450) made me do a little research and I think this might be an overall issue rather than just fingers itself. Spell queue which was introduced in tbc https://girlmeetswow.blogspot.com/2007/10/stopcasting-in-23-return-to-button.html . I believe this problem should be fixed once this is implemented. I had the idea this was added in cata but in cata they just added the ability to control to queue window rather than having a default value. I remember having to deal with the spell queue window interaction with instant casts to fully optimize boomkin gameplay in cata in order to make better use of eclipses and the behavior would make sense if you apply the same logic to fingers. You can read about it [here](https://web.archive.org/web/20110923151831/http://elitistjerks.com/f73/t110353-balance_cataclysm_4_2_a/#Advanced_Techniques) @ Advanced techniques section.
>Just tested on warmane pvp realm and fingers is fixed (wasn't back when i played there like a year ago) there and ghost charge does indeed work with the spell queue system implemented.
>Edit: You could maybe add a temporary/hackish fix and make it consume on-hit (while also making the cast that procs not consume its own proc) and it would behave mostly how it should behave btw. I don't see a spell queue system being added any time soon, not even accounting the amount of testing it would prob require as well.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/ccc9bb6679c232353939812760bedbd0f30b62cf

### Operating system

Ubuntu 20.04

### Custom changes or Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [mod-save-inventory](https://github.com/Nyeriah/mod-save-inventory)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### LINKED ISSUE #11876: [Mage] Fingers of Frost can proc on-cast and on-hit.
https://github.com/chromiecraft/chromiecraft/issues/3542

### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

### Current Behaviour

Fingers of Frost is currently proccing on-hit and on-cast, making it proc more often than it should.

### Expected Blizzlike Behaviour

Should only proc on cast only. It used to be a on-hit only trigger until patch 3.3. Was finding it weird how often it was proccing. 

### Source

>[Patch 3.3.0](https://wowpedia.fandom.com/wiki/Patch_3.3.0) (2009-12-08): This talent now triggers immediately on casting a spell rather than being delayed until the spell strikes the target. ([source](https://wowpedia.fandom.com/wiki/Fingers_of_Frost?oldid=2301949))

Also copy-pasting sources from another FoF issue:
https://www.youtube.com/watch?v=4X-GB9GB7Ic
https://www.youtube.com/watch?v=lr-ghl-CEL8
https://www.youtube.com/watch?v=8ET2Ct9_Ma8

### Steps to reproduce the problem

1. `.aura 44545`
2. cast Frostbolt (ID: 116) or Frostfire Bolt (ID: 44614) until you see a proc on-cast and a proc on-hit

### Extra Notes

Also see https://github.com/azerothcore/azerothcore-wotlk/issues/11338 and https://github.com/azerothcore/azerothcore-wotlk/issues/11339 for more FoF fun.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/7b4e1ef7d3d638f4ed182130af286a8f99528f53

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-dmf-switch](https://github.com/azerothcore/mod-dmf-switch)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna](https://github.com/azerothcore/mod-eluna)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-transmog](https://github.com/azerothcore/mod-transmog)
- [mod-weekend-xp](https://github.com/azerothcore/mod-weekend-xp)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Fixes #11338
Fixes #11876

## Issues Addressed:

- Closes #11338
- Closes #11876
- Closes https://github.com/chromiecraft/chromiecraft/issues/3123
- Closes https://github.com/chromiecraft/chromiecraft/issues/3542

## Tests Performed:

- Not tested.

## How to Test the Changes:

Combo Shatter
1. Go to a dummy with only Fingers of Frost specced,
2. When at only 1 charge of Fingers of Frost, attempt (and fail) to cast Deep Freeze.

Cast on hit proc
1. `.aura 44545`
2. cast Frostbolt (ID: 116) or Frostfire Bolt (ID: 44614) until you see a proc on-cast and a proc on-hit

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
- [X] Fingers of Frost can only proc on-cast and not on-hit
- [X] The ghost charge of FoF is working. You can cast Deep Freeze after the second Frostbolt.

Amazing. Works as intended from what I can see 👍

File: src/server/scripts/Spells/spell_mage.cpp [hunk 1]
```diff
@@ -932,6 +933,114 @@ class spell_mage_summon_water_elemental : public SpellScript
     }
 };
 
+#define FingersOfFrostScriptName "spell_mage_fingers_of_frost_proc_aura"
```
Comment (IntelligentQuantum) [inline src/server/scripts/Spells/spell_mage.cpp:936]: Isn't it better to use constexpr?

I said it only for my own knowledge xD

**Review by IntelligentQuantum — APPROVED** [review]

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -15831,18 +15832,35 @@ void Unit::ProcDamageAndSpellFor(bool isVictim, Unit* target, uint32 procFlag, u
...
+        if (!triggerData.aura->CallScriptCheckProcHandlers(itr->second, eventInfo))
+        {
+            continue;
+        }
+
+        // AuraScript Hook
+        if (!triggerData.aura->CallScriptCheckProcHandlers(itr->second, eventInfo))
+        {
             continue;
+        }
```
Comment (Nefertumm) [inline src/server/game/Entities/Unit/Unit.cpp:15845]: Why it is called two times?

File: src/server/scripts/Spells/spell_mage.cpp (hunk 1 above)
Comment (UltraNix) [reply src/server/scripts/Spells/spell_mage.cpp:936]: https://stackoverflow.com/questions/66176642/difference-between-constexpr-and-define
Short: Not here :)

**Pull request merged**